test/jruby/test_higher_javasupport.rb

```diff
@@ -1760,7 +1760,40 @@ def test_no_warnings_on_concurrent_class_const_initialization
     end
     # expect no already initialized constant warning written e.g.
     # ... warning: already initialized constant DefaultPackageClass
-    assert ! output.index('already initialized constant'), output
+    refute output.index('already initialized constant'), output
+  end
+
+  def test_no_warnings_on_concurrent_class_loading_with_include_package
+    mod = Module.new do
+      include_package 'java.lang'
+      include_package 'java.lang.annotation'
+      include_package 'java.lang.invoke'
+      include_package 'java.lang.management'
+      include_package 'java.lang.reflect'
+      include_package 'java.math'
+      include_package 'java.net'
+    end
+    output = with_stderr_captured do
+      threads = (0..100).map do
+        Thread.start do
+          sleep 0.01; mod::URLEncoder
+          sleep 0.01; mod::Field
+          sleep 0.01; mod::ThreadInfo
+          sleep 0.01; mod::LockInfo
+          sleep 0.01; mod::Array
```

Comment:
A better concurrency test could be to have each spawned thread wait on something like a CountDownLatch before trying to access each imported Java class, and have the main thread release the latches one at a time.

Reply:
yeah, that's a good point. although this actually does reproduce both types of warnings fairly reliably...